In [1]:
import pandas as pd

In [3]:
df = pd.read_csv("train.csv")

In [4]:
df

0    1    2    3    4    5         6         7  8    9  ...  \
0      236  488   16  221  382   97 -4.472136  0.107472  0  132  ...   
1      386  206  357  232    1  198  7.810250  0.763713  1  143  ...   
2      429   49  481  111  111  146  8.602325  0.651162  1  430  ...   
3      414  350  481  370  208  158  8.306624  0.424645  1  340  ...   
4      318  359   20  218  317  301  8.124038  0.767304  1  212  ...   
...    ...  ...  ...  ...  ...  ...       ...       ... ..  ...  ...   
89995  328   44  320  364    7   73  4.898979  0.563878  1  315  ...   
89996  217  472  174  327  255  389 -3.605551  0.861690  0  144  ...   
89997  189  406  213   57  494  190  6.164414  0.633704  1  221  ...   
89998  418  478  163   35  390   77 -3.605551  0.687309  0  289  ...   
89999  244  335  337  152  386  202  7.681146  0.613207  1   26  ...   

              44        45         46         47         48         49  \
0      13.340874  0.870542   1.962937   7.466666  11.547794   8.822916   
1      12.484882  7.168680   2.885415  12.413973  10.260494  10.091351   
2      14.030257  0.394970   8.160625  12.592059   8.937577   2.265191   
3       2.789577  6.416708  10.549814  11.456437   6.468099   2.519049   
4       1.886560  1.919999   2.268203   0.149421   4.105907  10.416291   
...          ...       ...        ...        ...        ...        ...   
89995   0.266362  6.185887   7.716837  10.144664   1.711649   3.849704   
89996   5.147059  4.209660  14.482770   1.375031   6.386263  10.107582   
89997  14.902807  2.886534   8.683168   4.522339   1.556421  10.971184   
89998  12.743029  2.525611  11.050145   6.589943  12.622192  10.596839   
89999  11.069077  6.303448  14.031393   1.877340  13.361607   2.164695   

              50         51         52     target  
0       9.046424   7.895535  11.010677  20.107472  
1       9.270888   3.173994  13.921871  61.763713  
2      11.255721  12.794841  12.080951  74.651162  
3       0.258284   9.317696   5.383098  69.424645  
4       6.816217   8.586960   4.512419  66.767304  
...          ...        ...        ...        ...  
89995  12.401903  14.195540   2.371207  24.563878  
89996  12.637902   6.576331   6.863238  13.861690  
89997   7.792226   8.422623   3.041409  38.633704  
89998   0.647584   8.746364   1.246682  13.687309  
89999  11.255181   3.404303   9.587379  59.613207  

[90000 rows x 54 columns]

so, there are 90000rows and 54column (53+1target). on the first side all are numerical. next, it is checked data types of columns and whether there are any empty rows/sells.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90000 entries, 0 to 89999
Data columns (total 54 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       90000 non-null  int64  
 1   1       90000 non-null  int64  
 2   2       90000 non-null  int64  
 3   3       90000 non-null  int64  
 4   4       90000 non-null  int64  
 5   5       90000 non-null  int64  
 6   6       90000 non-null  float64
 7   7       90000 non-null  float64
 8   8       90000 non-null  int64  
 9   9       90000 non-null  int64  
 10  10      90000 non-null  int64  
 11  11      90000 non-null  int64  
 12  12      90000 non-null  int64  
 13  13      90000 non-null  float64
 14  14      90000 non-null  float64
 15  15      90000 non-null  float64
 16  16      90000 non-null  float64
 17  17      90000 non-null  float64
 18  18      90000 non-null  float64
 19  19      90000 non-null  float64
 20  20      90000 non-null  float64
 21  21      90000 non-null  float64
 22

no empty columns and no categorical data. next, it is taken closer look on target data.

In [7]:
df['target'].min()

0.002634463140605

In [8]:
df["target"].max()

99.99948227942228

so, the range of target is [0, 100]. it is needed to predict one single number for every row in that range. it is regression task.

In [9]:
# all training data without target
X = df.drop('target', axis=1)
X

0    1    2    3    4    5         6         7  8    9  ...  \
0      236  488   16  221  382   97 -4.472136  0.107472  0  132  ...   
1      386  206  357  232    1  198  7.810250  0.763713  1  143  ...   
2      429   49  481  111  111  146  8.602325  0.651162  1  430  ...   
3      414  350  481  370  208  158  8.306624  0.424645  1  340  ...   
4      318  359   20  218  317  301  8.124038  0.767304  1  212  ...   
...    ...  ...  ...  ...  ...  ...       ...       ... ..  ...  ...   
89995  328   44  320  364    7   73  4.898979  0.563878  1  315  ...   
89996  217  472  174  327  255  389 -3.605551  0.861690  0  144  ...   
89997  189  406  213   57  494  190  6.164414  0.633704  1  221  ...   
89998  418  478  163   35  390   77 -3.605551  0.687309  0  289  ...   
89999  244  335  337  152  386  202  7.681146  0.613207  1   26  ...   

              43         44        45         46         47         48  \
0      10.125020  13.340874  0.870542   1.962937   7.466666  11.547794   
1      10.584802  12.484882  7.168680   2.885415  12.413973  10.260494   
2      10.273114  14.030257  0.394970   8.160625  12.592059   8.937577   
3       6.885179   2.789577  6.416708  10.549814  11.456437   6.468099   
4      11.162686   1.886560  1.919999   2.268203   0.149421   4.105907   
...          ...        ...       ...        ...        ...        ...   
89995   5.252173   0.266362  6.185887   7.716837  10.144664   1.711649   
89996  14.736929   5.147059  4.209660  14.482770   1.375031   6.386263   
89997   7.296410  14.902807  2.886534   8.683168   4.522339   1.556421   
89998  13.943818  12.743029  2.525611  11.050145   6.589943  12.622192   
89999   1.189056  11.069077  6.303448  14.031393   1.877340  13.361607   

              49         50         51         52  
0       8.822916   9.046424   7.895535  11.010677  
1      10.091351   9.270888   3.173994  13.921871  
2       2.265191  11.255721  12.794841  12.080951  
3       2.519049   0.258284   9.317696   5.383098  
4      10.416291   6.816217   8.586960   4.512419  
...          ...        ...        ...        ...  
89995   3.849704  12.401903  14.195540   2.371207  
89996  10.107582  12.637902   6.576331   6.863238  
89997  10.971184   7.792226   8.422623   3.041409  
89998  10.596839   0.647584   8.746364   1.246682  
89999   2.164695  11.255181   3.404303   9.587379  

[90000 rows x 53 columns]

In [10]:
y = df['target']
y

0        20.107472
1        61.763713
2        74.651162
3        69.424645
4        66.767304
           ...    
89995    24.563878
89996    13.861690
89997    38.633704
89998    13.687309
89999    59.613207
Name: target, Length: 90000, dtype: float64

first, it is better to normalise all columns but target(to not do back scaling after predictions). therefore, all 53 columns will be in one range -> [0, 1].  

In [12]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
# scaling only for X
X_scaled = scaler.fit_transform(X)

the hidden_test.csv file is given without ground truth, therefore to evaluate the model quality the train data is divided on train and test subsets.

In [62]:
from sklearn.model_selection import train_test_split

# 80% for training and 20% for testing
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=666)

as the origin of data is unknown as well as their meaning it is not possible to select columns that influence the most on the result. so, it is started from simple LinearRegression model to understand if is needed to use smth more complex.

In [32]:
from sklearn import linear_model
from sklearn.metrics import root_mean_squared_error

linear = linear_model.LinearRegression().fit(X_train, y_train)
y_pred = linear.predict(X_test)
root_mean_squared_error(y_test, y_pred)

28.705324172973633

not good results as it is already known that the target range is [0, 100] the rmse 28 means that on average prediction deviates by +-28. it is quite big for that range

as it is said in the instruction file it is needed to provide training file, so, it is probably expected model to be trained. therefore, next code is simple regression model with fully connected layers and relu activation func (to add non-linearity in model, otherwise it is just linear transformation regardless of the number of layers)

In [92]:
import torch
from torch import nn, optim

class RegressionModel(nn.Module):
    def __init__(self, input_size):
        super(RegressionModel, self).__init__()

        self.layers = nn.Sequential(
            nn.Linear(input_size, 256),
            nn.ReLU(),
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, 32),
            nn.ReLU(),
            nn.Linear(32, 1),
            # nn.ReLU(),
            # nn.Linear(16, 1)
        )

    def forward(self, x):
        return self.layers(x)

to start training it is needed to transform data into tensors and create dataloaders

In [65]:
# as y is still in pandas type Serias it is needed to be transform into numpy array
y_train_np = y_train.values.reshape(-1, 1)
y_test_np = y_test.values.reshape(-1, 1)
type(y_train_np)

numpy.ndarray

In [66]:
from torch.utils.data import Dataset, DataLoader

class RegressionDataset(Dataset):
    def __init__(self, X_data, y_data):
        self.X_data = torch.tensor(X_data, dtype=torch.float32)
        self.y_data = torch.tensor(y_data, dtype=torch.float32)

    def __len__(self):
        return len(self.X_data)

    def __getitem__(self, index):
        return self.X_data[index], self.y_data[index]

In [69]:
train_dataset = RegressionDataset(X_train, y_train_np)
test_dataset = RegressionDataset(X_test, y_test_np)
train_dataset

In [89]:
batch_size = 64

train_loader = DataLoader(
    dataset=train_dataset,
    batch_size=batch_size,
    # after each epoch data is shuffled
    shuffle=True
)

test_loader = DataLoader(
    dataset=test_dataset,
    batch_size=batch_size,
    shuffle=False
)

In [101]:
# X_train.shape[1] is number of columms == input size of model
reg_model = RegressionModel(X_train.shape[1])

# loss func for regression mean squared error
criterion = nn.MSELoss()

# optimiser with learning rate
optimiser = optim.Adam(reg_model.parameters(), lr=0.0001)

In [102]:
epochs = 50
for epoch in range(epochs):
    reg_model.train()
    for X_batch, y_batch in train_loader:
        y_pred = reg_model(X_batch)
        # now it is rmse loss
        loss = torch.sqrt(criterion(y_pred, y_batch))
        optimiser.zero_grad()
        loss.backward()
        optimiser.step()

    if (epoch + 1) % 5 == 0:
        print(f'Epoch [{epoch + 1}/{epochs}], Loss RMSE: {loss.item():.6f}')

Epoch [5/50], Loss RMSE: 9.639120
Epoch [10/50], Loss RMSE: 1.455069
Epoch [15/50], Loss RMSE: 0.557094
Epoch [20/50], Loss RMSE: 0.354092
Epoch [25/50], Loss RMSE: 0.509757
Epoch [30/50], Loss RMSE: 0.377874
Epoch [35/50], Loss RMSE: 0.252047
Epoch [40/50], Loss RMSE: 0.432672
Epoch [45/50], Loss RMSE: 0.497123
Epoch [50/50], Loss RMSE: 0.330337


In [103]:
import numpy as np

reg_model.eval()
all_y_true = []
all_y_pred = []
with torch.no_grad():
    for X_batch, y_batch in test_loader:
        y_pred = reg_model(X_batch)
        all_y_true.append(y_batch.numpy())
        all_y_pred.append(y_pred.numpy())

y_test_final = np.concatenate(all_y_true)
y_pred_final = np.concatenate(all_y_pred)

rmse = root_mean_squared_error(y_test_final, y_pred_final)
print(f"Test RMSE : {rmse:.4f}")

Test RMSE : 0.2204


it is decided to stop here because I believe that rmse 0.22 is good for the range [0, 1]. For train.py file it is gonna used all train data without train/test devision.